sdk/typescript/runtime/main.go

```diff
@@ -340,17 +375,138 @@ func (t *TypescriptSdk) detectRuntime(ctx context.Context, modSource *ModuleSour
+
+	// read contents of package.json
+	json, err := source.File("package.json").Contents(ctx)
+	if err == nil {
+		// Check if the dagger.runtime field is set to "bun" first
+		value := gjson.Get(json, "dagger.runtime").String()
+		if value != "" && strings.Contains(value, "bun") {
+			return BunManager, "", nil
+		}
+
+		value = gjson.Get(json, "packageManager").String()
+		if value != "" {
+			// Retrieve the package manager and version from the value (e.g., yarn@4.2.0, pnpm@8.5.1)
+			packageManager, version, _ := strings.Cut(value, "@")
+
+			switch SupportedPackageManager(packageManager) {
+			case Yarn, Pnpm, Npm:
+				return SupportedPackageManager(packageManager), version, nil
+			default:
+				return "", "", fmt.Errorf("detected unknown package manager: %s", packageManager)
+			}
+		}
+	}
+
+	// Try to detect package manager from lock files
+	entries, err := source.Entries(ctx)
+	if err == nil {
+		if slices.Contains(entries, "bun.lockb") {
+			return BunManager, "", nil
+		}
+
+		if slices.Contains(entries, "package-lock.json") {
+			return Npm, "", nil
+		}
+
+		if slices.Contains(entries, "yarn.lock") {
+			return Yarn, "", nil
+		}
+
+		if slices.Contains(entries, "pnpm-lock.yaml") {
+			return Pnpm, "", nil
+		}
+	}
+
+	return Yarn, YarnDefaultVersion, nil
+}
+
+// generateLockFile generate a lock file for the matching package manager.
+func generateLockFile(ctx context.Context, ctr *Container, packageManager SupportedPackageManager, version string) (*Container, error) {
+	switch packageManager {
+	case Yarn:
+		// Sadly, yarn < v3 doesn't support generating a lockfile without installing the dependencies.
+		// So we use npm to generate the lockfile and then import it into yarn.
+		return ctr.
+			WithExec([]string{"corepack", "enable"}).
+			WithExec([]string{"corepack", "use", fmt.Sprintf("yarn@%s", version)}).
+			WithExec([]string{"yarn", "install", "--mode", "update-lockfile"}), nil
+	case Pnpm:
+		return ctr.
+			WithExec([]string{"npm", "install", "-g", fmt.Sprintf("pnpm@%s", version)}).
+			WithExec([]string{"pnpm", "install", "--lockfile-only"}), nil
+	case Npm:
+		return ctr.
+			WithExec([]string{"npm", "install", "-g", fmt.Sprintf("npm@%s", version)}).
+			WithExec([]string{"npm", "install", "--package-lock-only"}), nil
+	case BunManager:
+		return ctr.
+			WithExec([]string{"bun", "install", "--no-verify", "--no-progress"}), nil
+	default:
+		return nil, fmt.Errorf("detected unknown package manager: %s", packageManager)
+	}
+}
+
+// installDependencies installs the dependencies using the detected package manager.
+//
+// If the package.json file has a field "packageManager", it will use that to
+// determine the package manager to use. Otherwise, it will use the default
+// package manager based on the lock files present in the module.
+//
+// If none of the above works, it will use yarn.
+//
+// Except if the package.json has an invalid value in field "packageManager", this
+// // function should never return an error.
```

Comment:
Looks like this is duplicated from `detectPackageManager`. Doesn't make sense here.